### PR TITLE
rustPlatform.buildRustPackage: build auditable binaries

### LINKED ISSUE #200563: Support for building auditable rust binaries with cargo-auditable
[cargo-auditable](https://github.com/rust-secure-code/cargo-auditable) is a tool to store the exact versions of dependencies in rust binaries and makes it easier/possible to audit the dependency tree of rust binaries. It was added to nixpkgs in #191821, and there is currently an open RFC to implement this within cargo: (https://github.com/rust-lang/rfcs/pull/2801)

Void linux recently integrated this with its build system: https://github.com/void-linux/void-packages/pull/40272, and thank you @jcgruenhage for the suggestion

Things to be done:
- implementation (here's two I was able to think of):
  - `rustPlatformAuditable` that uses `cargo-auditable` instead of `cargo` directly
  - an `auditable` option in `buildRustPackage` and `buildRustCrate`
- Bootstrap cargo-auditable
- Figure out whether we want to make this the default

### PR DESCRIPTION
**Pull request by figsoda**
###### Description of changes

closes #200563 

bootstraps `cargo-auditable` and makes all packages built by `buildRustPackage` auditable unless overridden with `auditable = false`

built `fd` on master and ran `rust-audit-info` on it, it should have no sizable impact on binary size

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
Any other distros doing this?

**Comment by figsoda** [comment]
just void right now as far as I'm aware

**Comment by zowoq** [comment]
> https://github.com/NixOS/nixpkgs/issues/200563
> Figure out whether we want to make this the default

TBH I think making it the default is a bit premature. Keep it opt-in for testing and revisit making it the default has been in master for a few weeks?

**Comment by lukegb** [comment]
Given that the changes to the hook will cause every package to be rebuilt anyway I don't think there's necessarily much harm in enabling it by default now, but I'm also not familiar enough with the Rust ecosystem to know whether or not just doing this is likely to be harmful in some edgecase.

**Comment by zowoq** [comment]
> maybe we can opt in for a small set of packages and make it the default when it's observed to be stable?

^ This is what I meant when I said this:

> Keep it opt-in for testing and revisit making it the default has been in master for a few weeks?

___

> Given that the changes to the hook will cause every package to be rebuilt anyway

Don't see how this makes any difference as packages will be rebuilt in the following staging cycles anyway?

> I don't think there's necessarily much harm in enabling it by default now, but I'm also not familiar enough with the Rust ecosystem to know whether or not just doing this is likely to be harmful in some edgecase.

Reverting it after it is in master would mean another staging cycle so I think a change that affects every use of `buildRustPackage` warrants a bit of testing before making it the default?

**Comment by lukegb** [comment]
I think my main complaint is that "a bit of testing" does not a testing plan make :)

I agree that it is probably a good idea to do some form of testing, but also... this _is_ nixos-unstable. The purest form of smoke testing here would be "running all the NixOS tests that exist and seeing if they pass", which is a good bet to happen in staging-next, and if there are any obvious problems there it's easy to back out. I agree if it escapes into master then it's a bit more annoying to revert but at least we'll have evidence of real problems at that point.

What sort of shape of testing would you be interested in to help build confidence in this change?

**Comment by zowoq** [comment]
> What sort of shape of testing would you be interested in to help build confidence in this change?

This sounds good:

> maybe we can opt in for a small set of packages and make it the default when it's observed to be stable

**Comment by figsoda** [comment]
changed `auditable = false` to the default and made a couple popular programs auditable also added jumpy for some diversity

**Comment by zowoq** [comment]
Maybe `mdbook` as well as it's in the `nix` build closure?

File: pkgs/tools/video/rav1e/default.nix
Comment (zowoq) [inline pkgs/tools/video/rav1e/default.nix:24]: Doesn't need to happen in this PR but any way we could avoid needing to do this? i.e wrap `cargo-auditable` as `cargo` or something like that?

Comment (figsoda) [reply pkgs/tools/video/rav1e/default.nix:24]: How do we stop the wrapper from recursively calling itself? Looks like cargo-auditable just calls `cargo` directly

https://github.com/rust-secure-code/cargo-auditable/blob/6275e86f503d9419b1528aa38af128882a648ba4/cargo-auditable/src/cargo_auditable.rs#L38

Comment (figsoda) [reply pkgs/tools/video/rav1e/default.nix:24]: Maybe we can look at `CARGO_AUDITABLE_ORIG_ARGS`, but I'm not sure if we should rely on that since it is internal and undocumented

Comment (figsoda) [reply pkgs/tools/video/rav1e/default.nix:24]: Another way to do it is just to add something like a `cargoAuditableHook` that sets `RUSTC_WORKSPACE_WRAPPER`

Comment (zowoq) [reply pkgs/tools/video/rav1e/default.nix:24]: > How do we stop the wrapper from recursively calling itself? Looks like cargo-auditable just calls `cargo` directly

Make `cargo` into `cargo-unwrapped`, substitute `cargo` for `cargo-unwrapped` in `cargo-auditable` and then wrap it?

Comment (figsoda) [reply pkgs/tools/video/rav1e/default.nix:24]: > Make `cargo` into `cargo-unwrapped`

Sounds like the most future-proof way of doing it, but it would complicate the bootstrap process a bit.

Due to the complexity of this, I think it is worth a separate issue once this is merged

Comment (zowoq) [reply pkgs/tools/video/rav1e/default.nix:24]: > Due to the complexity of this, I think it is worth a separate issue once this is merged

That's fine but I do think it's something that should be resolved before `auditable` is made the default.

`$cargoCommand` in an "extra" step like it is here isn't too bad but much easier for everyone if people don't need to know about it.

Should also mean that it works without changes for packages that have cmake, make, meson, etc build systems calling `cargo`.

Comment (Shnatsel) [reply pkgs/tools/video/rav1e/default.nix:24]: Hey! `cargo auditable` author here 👋 

I'll see what I can do about providing a robust way of using `cargo auditable` as a drop-in replacement for `cargo.`

Here's the upstream issue so you can track the progress: https://github.com/rust-secure-code/cargo-auditable/issues/89

**Comment by jcgruenhage** [comment]
> Any other distros doing this? Besides void?

Also not aware of other distros, but we've been building any update to Rust packages that use the cargo build style with auditable, and we have not run into problems with it.

Are there any other distros where adopting this would be easy to implement? I've looked at both Arch and Alpine, and they don't abstract away the build commands into a build style or hook it something like that, meaning that adopting that for all packages is a way bigger change than it would be for Void or NixOS.

File: pkgs/tools/video/rav1e/default.nix
Comment (figsoda) [reply pkgs/tools/video/rav1e/default.nix]: @Shnatsel thanks for all the work on cargo-auditable.
It works! I wasn't able to use an alias since `cargoBuildHook` uses `env`, which doesn't pick up aliases, but a shell script wrapper does the job

**Comment by figsoda** [comment]
I added `cargo-auditable-cargo-wrapper` and based `buildRustPackage` on it. Not sure about the name, also considered `cargo-auditable.passthru.cargoWrapper` but didn't do that for the simplicity of this initial implementation. open to suggestions

Made `libsrvg` auditable as per @Shnatsel's suggestion since now it's possible with the wrapper

It should be simple to add auditable support for `buildRustCrate` as well, but I am not familiar with it so I'll leave it to someone else

**Comment by Shnatsel** [comment]
I've released `cargo auditable` v0.6.0 with support for using it as a drop-in replacement for `cargo`, so that you don't have to use a development snapshot.

**Comment by Shnatsel** [comment]
To clarify the potential failure modes: If something goes wrong, the build should fail. It is highly unlikely that a build would produce a non-functional binary, since the data injection mechanism is lifted directly from the Rust compiler, so it's well-tested. And it simply emits an additional object file for the linker to process, so it simply reuses through all the usual, well-trodden codepaths.

**Comment by figsoda** [comment]
@zowoq is this good to merge?

File: pkgs/development/libraries/librsvg/default.nix
Comment (zowoq) [inline pkgs/development/libraries/librsvg/default.nix:18]: Add a TODO here as well for when auditable is the default?

Comment (figsoda) [reply pkgs/development/libraries/librsvg/default.nix:18]: Currently it only works for `buildRustPackage`, do you think it should just be a override in cargo then?

Comment (zowoq) [reply pkgs/development/libraries/librsvg/default.nix:18]: Making it consistent (opt-out) tree-wide would be better but I guess after `buildRustPackage` is covered there isn't that much left.

Leave it for now and we can revisit this when auditable is made the default?

Comment (figsoda) [reply pkgs/development/libraries/librsvg/default.nix:18]: sounds good

**Review by zowoq — APPROVED** [review]

**Comment by Mindavi** [comment]
When auditable is set to true, this causes a regression when cross-compiling rust packages where a cross-compiled rustc and cargo end up in the closure (causing failures since that doesn't work yet):

```
rick@nixos-asus:~/nixpkgs$ nom build -f default.nix pkgsCross.aarch64-multiplatform.ripgrep -L
these 12 derivations will be built:
  /nix/store/zk6as569vzji8979214qig7y0y26isyr-cargo-1.66.1.drv
  /nix/store/4zqkw80n5lnc9lg4hngqqxnwwnhx3gva-rustc-aarch64-unknown-linux-gnu-1.66.1.drv
  /nix/store/7v21vgng6qjpg4rw96x8nz8b87snv7nm-cargo-check-hook.sh.drv
  /nix/store/jahm46k3xf68zkf15gvjgpmlm2jjrpv2-cargo-build-hook.sh.drv
  /nix/store/9rhmncrcsza15fii2f6644n4mbrhkvbf-cargo-aarch64-unknown-linux-gnu-1.66.1.drv
  /nix/store/hl8g9lskjiqb30sdml2acnn8lja88rqf-cargo-build-hook.sh.drv
  /nix/store/zs37bfg93h0879lglasm7fd96c9v54nm-cargo-check-hook.sh.drv
  /nix/store/30wx8md8q7wlrkqln6ngd6cskrbw55d4-cargo-auditable-aarch64-unknown-linux-gnu-0.6.0.drv
  /nix/store/csz9pipmxk9p9nwcmarn4fz3d4b9vans-cargo.drv
  /nix/store/9cj5wpz3k9lna0vcr2rfs0vwhc31fwpw-cargo-auditable-aarch64-unknown-linux-gnu-0.6.0.drv
  /nix/store/18vrjsy9fsljw6nv0xzpdn083zfzm9dm-cargo.drv
  /nix/store/3h6lcrmxl2scm42bvr34dsdjbg3nw9dn-ripgrep-aarch64-unknown-linux-gnu-13.0.0.drv
┏━ Dependency Graph:
┃             ┌─ ⏳︎ cargo-check-hook.sh waiting for 1 ⏵︎
┃             │     ┌─ ⏳︎ cargo-build-hook.sh waiting for 1 ⏵︎
┃             │     ├─ ⏳︎ cargo-check-hook.sh waiting for 1 ⏵︎
┃             │     │  ┌─ ⏵︎ cargo-1.66.1 (buildPhase) ⏱︎ 15s (∅︎ 6m42s)
┃             │     ├─ ⏳︎ rustc-aarch64-unknown-linux-gnu-1.66.1
┃             │  ┌─ ⏳︎ cargo-aarch64-unknown-linux-gnu-1.66.1
┃             ├─ ⏳︎ cargo-build-hook.sh
┃          ┌─ ⏳︎ cargo-auditable-aarch64-unknown-linux-gnu-0.6.0
┃       ┌─ ⏳︎ cargo
┃    ┌─ ⏳︎ cargo-auditable-aarch64-unknown-linux-gnu-0.6.0
┃ ┌─ ⏳︎ cargo
┃ ⏳︎ ripgrep-aarch64-unknown-linux-gnu-13.0.0
┣━━━ Builds           
┗━ ∑︎ ⏵︎ 1 │ ✔︎ 0 │ ⏳︎ 11 │ ⏱︎ 22s^C
```

**Comment by figsoda** [comment]
https://github.com/NixOS/nixpkgs/pull/210394

**Pull request merged**
